File: src/types/examples.ts
/**
 * 类型使用示例
 * 这个文件展示如何使用连接类型
 */

import type {
  Connection,
  Group,
  SSHConfig,
  DatabaseConfig,
  ConnectionSession,
} from './index'

// 示例：创建 SSH 连接
function createSSHConnection(): Connection {
  return {
    id: '1',
    name: 'Production Server',
    type: 'ssh',
    group_id: 'group-1',
    config: {
      host: '192.168.1.100',
      port: 22,
      username: 'admin',
      auth_method: 'password',
      password: 'password123',
    },
    created_at: '2025-02-05T00:00:00Z',
    updated_at: '2025-02-05T00:00:00Z',
  }
}

// 示例：创建 MySQL 连接
function createMySQLConnection(): Connection {
  return {
    id: '2',
    name: 'Production Database',
    type: 'mysql',
    config: {
      host: 'localhost',
      port: 3306,
      username: 'root',
      password: 'password',
      database: 'mydb',
      ssl: true,
    },
    created_at: '2025-02-05T00:00:00Z',
    updated_at: '2025-02-05T00:00:00Z',
  }
}

// 示例：创建带跳板机的 SSH 连接
function createSSHConnectionWithJumpHost(): Connection {
  return {
    id: '3',
    name: 'Bastion Server',
    type: 'ssh',
    config: {
      host: '10.0.0.1',
      port: 22,
      username: 'admin',
      auth_method: 'key',
      private_key_path: '/home/user/.ssh/id_rsa',
      passphrase: 'keypass',
      jump_host: {
        host: 'bastion.example.com',
        port: 22,
        username: 'bastion_user',
        auth_method: 'password',
        password: 'bastion_pass',
      },
    },
    created_at: '2025-02-05T00:00:00Z',
    updated_at: '2025-02-05T00:00:00Z',
  }
}

// 示例：创建分组
function createGroup(): Group {
  return {
    id: 'group-1',
    name: 'Production Servers',
    parent_id: undefined,
    icon: 'folder',
    created_at: '2025-02-05T00:00:00Z',
    updated_at: '2025-02-05T00:00:00Z',
  }
}

// 示例：创建连接会话
function createConnectionSession(): ConnectionSession {
  return {
    id: 'session-1',
    connection_id: '1',
    status: 'connected',
    connected_at: '2025-02-05T00:00:00Z',
  }
}

// 示例：类型守卫函数
function isSSHConfig(config: SSHConfig | DatabaseConfig): config is SSHConfig {
  return 'auth_method' in config
}

function isDatabaseConfig(
  config: SSHConfig | DatabaseConfig,
): config is DatabaseConfig {
  return 'ssl' in config
}

// 示例：使用类型守卫
function processConnection(connection: Connection) {
  if (isSSHConfig(connection.config)) {
    console.log(`SSH Connection: ${connection.config.host}:${connection.config.port}`)
  } else if (isDatabaseConfig(connection.config)) {
    console.log(
      `Database Connection: ${connection.config.host}:${connection.config.port}/${connection.config.database}`,
    )
  }
}

// 导出示例
export {
  createGroup,
  createMySQLConnection,
  createSSHConnection,
  createSSHConnectionWithJumpHost,
  createConnectionSession,
  isSSHConfig,
  isDatabaseConfig,
  processConnection,
}

File: src-tauri/src/models/examples.rs
// 模块使用示例文件
// 这个文件展示了如何使用我们定义的类型
// 在实际实现 commands 时会被使用

use crate::models::connection::*;

#[allow(dead_code)]
fn example_create_ssh_connection() -> Connection {
    Connection {
        id: "1".to_string(),
        name: "Production Server".to_string(),
        connection_type: ConnectionType::Ssh,
        group_id: Some("group-1".to_string()),
        config: Config::Ssh(SSHConfig {
            host: "192.168.1.100".to_string(),
            port: 22,
            username: "admin".to_string(),
            auth_method: AuthMethod::Password,
            password: Some("password123".to_string()),
            private_key_path: None,
            passphrase: None,
            jump_host: None,
        }),
        created_at: "2025-02-05T00:00:00Z".to_string(),
        updated_at: "2025-02-05T00:00:00Z".to_string(),
    }
}

#[allow(dead_code)]
fn example_create_mysql_connection() -> Connection {
    Connection {
        id: "2".to_string(),
        name: "Production Database".to_string(),
        connection_type: ConnectionType::Mysql,
        group_id: None,
        config: Config::Database(DatabaseConfig {
            host: "localhost".to_string(),
            port: 3306,
            username: "root".to_string(),
            password: "password".to_string(),
            database: Some("mydb".to_string()),
            ssl: Some(true),
        }),
        created_at: "2025-02-05T00:00:00Z".to_string(),
        updated_at: "2025-02-05T00:00:00Z".to_string(),
    }
}

#[allow(dead_code)]
fn example_create_group() -> Group {
    Group {
        id: "group-1".to_string(),
        name: "Production Servers".to_string(),
        parent_id: None,
        icon: Some("folder".to_string()),
        created_at: Some("2025-02-05T00:00:00Z".to_string()),
        updated_at: Some("2025-02-05T00:00:00Z".to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ssh_connection_serialization() {
        let conn = example_create_ssh_connection();
        let json = serde_json::to_string(&conn).unwrap();

        // 验证序列化结果
        assert!(json.contains("\"type\":\"ssh\""));
        assert!(json.contains("192.168.1.100"));
        assert!(json.contains("admin"));

        // 反序列化测试
        let _deserialized: Connection = serde_json::from_str(&json).unwrap();
    }

    #[test]
    fn test_mysql_connection_serialization() {
        let conn = example_create_mysql_connection();
        let json = serde_json::to_string(&conn).unwrap();

        // 验证序列化结果
        assert!(json.contains("\"type\":\"mysql\""));
        assert!(json.contains("localhost"));
        assert!(json.contains("3306"));

        // 反序列化测试
        let _deserialized: Connection = serde_json::from_str(&json).unwrap();
    }

    #[test]
    fn test_group_serialization() {
        let group = example_create_group();
        let json = serde_json::to_string(&group).unwrap();

        // 验证序列化结果
        assert!(json.contains("Production Servers"));

        // 反序列化测试
        let _deserialized: Group = serde_json::from_str(&json).unwrap();
    }
}
